File: libefb-swift/Sources/LibEFB/Airspace.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2024 Joe Pearson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import CoreLocation
import EFB
import Foundation

extension CLLocationCoordinate2D {
    init(_ coordinate: EFB.Coordinate) {
        self.init(
            latitude: Double(coordinate.latitude),
            longitude: Double(coordinate.longitude)
        )
    }
}

public enum VerticalDistance {
    case agl(Int)
    case altitude(Int)
    case fl(Int)
    case gnd
    case msl(Int)
    case unlimited
}

extension VerticalDistance {
    init(_ value: EFB.VerticalDistance) {
        switch value.tag {
        case EFB.Agl:
            self = .agl(Int(value.agl))
        case EFB.Altitude:
            self = .altitude(Int(value.altitude))
        case EFB.Fl:
            self = .fl(Int(value.fl))
        case EFB.Gnd:
            self = .gnd
        case EFB.Msl:
            self = .msl(Int(value.msl))
        case EFB.Unlimited:
            self = .unlimited
        default:
            self = .unlimited
        }
    }
}

public struct Airspace: Identifiable {
    public let id = UUID()
    public let name: String
    public let classification: AirspaceClassification
    public let coordinates: [CLLocationCoordinate2D]
    public let ceiling: VerticalDistance
    public let floor: VerticalDistance
    public let annotation: AirspaceAnnotation

    init(_ airspace: EFB.Airspace) {
        let polygon = Array(
            UnsafeBufferPointer(
                start: airspace.polygon.data,
                count: airspace.polygon.len))

        self.name = String(cString: airspace.name)
        self.classification = airspace.classification

        self.coordinates = polygon.map { coordinate in
            CLLocationCoordinate2D.init(coordinate)
        }

        self.ceiling = VerticalDistance.init(airspace.ceiling)
        self.floor = VerticalDistance.init(airspace.floor)
        self.annotation = airspace_annotation(airspace.classification)

        airspace_free(airspace)
    }
}

File: efb/src/nd/airspace.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2024 Joe Pearson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::fmt::{Display, Formatter, Result};

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

use crate::geom::Polygon;
use crate::VerticalDistance;

pub type Airspaces = Vec<Airspace>;

/// Airspace.
///
/// The airspace is classified by the `class` and enclosed by the `polygon`.
/// It ranges from the `floor` to `ceiling` in vertical direction.
#[repr(C)]
#[derive(Clone, PartialEq, Debug)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Airspace {
    pub name: String,
    pub class: AirspaceClass,
    pub ceiling: VerticalDistance,
    pub floor: VerticalDistance,
    pub polygon: Polygon,
}

#[repr(C)]
#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum AirspaceClass {
    A,
    B,
    C,
    D,
    E,
    F,
    G,
    CTR,
    Restricted,
    Danger,
    Prohibited,
}

impl Display for AirspaceClass {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            AirspaceClass::A => write!(f, "Class A"),
            AirspaceClass::B => write!(f, "Class B"),
            AirspaceClass::C => write!(f, "Class C"),
            AirspaceClass::D => write!(f, "Class D"),
            AirspaceClass::E => write!(f, "Class E"),
            AirspaceClass::F => write!(f, "Class F"),
            AirspaceClass::G => write!(f, "Class G"),
            AirspaceClass::CTR => write!(f, "CTR"),
            AirspaceClass::Restricted => write!(f, "Restricted"),
            AirspaceClass::Danger => write!(f, "Danger"),
            AirspaceClass::Prohibited => write!(f, "Prohibited"),
        }
    }
}

impl Display for Airspace {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(
            f,
            "{}: {} | {}/{}",
            self.name, self.class, self.ceiling, self.floor
        )
    }
}
